src/Common/src/System/Data/ProviderBase/DbConnectionPoolGroup.cs

```diff
@@ -31,7 +31,7 @@ sealed internal class DbConnectionPoolGroup
         private readonly DbConnectionOptions _connectionOptions;
         private readonly DbConnectionPoolKey _poolKey;
         private readonly DbConnectionPoolGroupOptions _poolGroupOptions;
-        private ConcurrentDictionary<DbConnectionPoolIdentity, DbConnectionPool> _poolCollection;
```

Comment:
At least this field isn't set via double-checked locking (it's behind a single explicit lock, or in the constructor), so not relevant.

Reply:
The potential race is between line 122 and line 165. But I agree, that `volatile` isn't going to help with the reads on lines 165/170, which is what the query actually found.

Reply:
Ah, you're right, I should have been locking at more of the reads.

Reply:
This is going to add a bunch of unnecessary volatile reads and writes, as there are lots of accesses where it's not actually necessary (e.g. all of the reads while holding the lock don't need to be volatile).  I'd rather see this one done with Volatile.Read/Write on just the accesses where it's needed.  This is unlikely to make much of a difference on x86-64, but it could on ARM.